### PR TITLE
Support for string slices in contracts/predicates/scripts

### PR DESCRIPTION
**Pull request by xunilrj**
## Description

This PR, for encoding v1, allows string slices in inputs and outputs of scripts and contracts; and in inputs of predicates.
For encoding v0, the limitation is kept.

This PR will also land the necessary start of future improvements on "slices" (including string slices).

This PR closes https://github.com/FuelLabs/sway/issues/5145

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-core/src/language/ty/expression/contract.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@ use crate::language::ty::*;
 
 #[derive(Clone, Debug)]
 pub struct ContractCallParams {
-    pub(crate) func_selector: [u8; 4],
+    pub(crate) func_selector: Option<[u8; 4]>,
```
Comment (IGI-111) [inline sway-core/src/language/ty/expression/contract.rs:5]: I think this needs a comment to explain what the type means for the `None` case.

File: sway-core/src/type_system/info.rs [hunk 2]
```diff
@@ -1020,6 +1020,7 @@ impl TypeInfo {
             // TODO-IG: No references in ABIs according to the RFC. Or we want to have them?
             // TODO-IG: Depending on that, we need to handle `Ref` here as well.
             _ => {
+                println!("{}", std::backtrace::Backtrace::force_capture());
```
Comment (IGI-111) [inline sway-core/src/type_system/info.rs:1023]: This looks like a debug print that shouldn't be included.

**Review by IGI-111 — COMMENTED** [review]
A few nits but overall this is cleaner than I would expect.

Can't wait to be able to ditch v0 and simplify this even more.

File: sway-core/src/type_system/info.rs (hunk 2 above)
Comment (xunilrj) [reply sway-core/src/type_system/info.rs:1023]: Done.

File: sway-core/src/language/ty/expression/contract.rs (hunk 1 above)
Comment (xunilrj) [reply sway-core/src/language/ty/expression/contract.rs:5]: Done.

**Review by IGI-111 — DISMISSED** [review]

**Review by IGI-111 — DISMISSED** [review]

**Review by IGI-111 — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**